Pre-Built SIPNET Binary for GitHub Actions

AritraDey-Dev opened the pull request:
Fixes #3427

Add CI workflow to build and upload SIPNET binary as an artifact and update PEcAn workflow to use pre-built SIPNET binary.

* **ci.yml**
  - Add steps to build SIPNET binary.
  - Add steps to upload SIPNET binary as an artifact.

* **sipnet.yml**
  - Remove steps to check out and build SIPNET from source.
  - Add steps to download pre-built SIPNET binary.
  - Add steps to use pre-built SIPNET binary.

---

infotroph reviewed (changes requested):
Thanks for working on this! To meet the requirements of #3427 you'll need to make some changes:

* Building the binary should be done in the [sipnet repo](https://github.com/pecanproject/sipnet). Determining how to get the built binary from there to here is the crux of the task.
* In all repositories Github Actions workflows have to be in `.github/workflows`, not in the project root.
* The changes to PEcAn's CI should modify existing workflows rather than add new ones.

AritraDey-Dev commented:
> Thanks for working on this! To meet the requirements of #3427 you'll need to make some changes:
> 
> * Building the binary should be done in the [sipnet repo](https://github.com/pecanproject/sipnet). Determining how to get the built binary from there to here is the crux of the task.
> * In all repositories Github Actions workflows have to be in `.github/workflows`, not in the project root.
> * The changes to PEcAn's CI should modify existing workflows rather than add new ones.

Sure I will look into this changes.

AritraDey-Dev commented:
i have made the required changes.can you please look into those.

dlebauer commented:
@AritraDey-Dev Looks good to me! I am not sure if Lines 61-73 of ci.yml need to be there, now that the binary building will be done elsewhere. 

I put a note in https://github.com/PecanProject/sipnet/issues/27 and will update the curl statements once we have a release with binaries.

infotroph commented on .github/workflows/ci.yml:
```diff
@@ -58,7 +58,16 @@ jobs:
+        uses: actions/upload-artifact@v2
+        with:
+          name: sipnet-binary
+          path: ./bin/sipnet 
```
No changes are needed in this file -- everything related to getting Sipnet installed is encapsulated inside sipnet.yml and nothing should change in how we call it.

AritraDey-Dev replied:
i have done the modifications.

AritraDey-Dev commented:
@infotroph i have Updated upload-artifact to v3. SIPNET installation remains encapsulated in sipnet.yml, so no further modifications were needed. Let me know if anything else is required!

allgandalf commented on .github/workflows/ci.yml:
```diff
@@ -61,4 +61,4 @@ jobs:
     with:
       R-version: ${{ matrix.R }}
-    secrets: inherit
+    secrets: inherit
```
I don't understand this code diff, what has changed here? did you add a extra space or something ? if so why?

AritraDey-Dev replied:
that is mistake from my side while changing the sipnet file and testing it.Thanks for pointing out this.

AritraDey-Dev replied:
i am also not sure,may be while coding there is some unnecessary spaces that came.I will revert it.

allgandalf commented on .github/workflows/sipnet.yml:
```diff
@@ -49,16 +49,15 @@ jobs:
       run: |
-        cd ${GITHUB_WORKSPACE}/sipnet
-        make sipnet
+        curl -L -o sipnet-linux "https://drive.google.com/uc?export=download&id=1sRTFfs9Z9osmMrtLlwVTqxeUDxgVlrEv"
```
@dlebauer @infotroph are we sure that this drive link will always be live for our project ? joining mid way so do not have the best context

dlebauer replied:
That was just a placeholder until I got builds posted with different releases, the latest is 1.3.0 https://github.com/PecanProject/sipnet/releases/tag/v1.3.0

This will always resolve to the latest release https://github.com/PecanProject/sipnet/releases/latest but the version number in the binary name will change.

dlebauer replied:
This should be the Linux build https://github.com/PecanProject/sipnet/releases/download/v1.3.0/sipnet-linux-v1.3.0

allgandalf commented on .github/workflows/sipnet.yml:
```diff
@@ -49,16 +49,15 @@ jobs:
-        cd ${GITHUB_WORKSPACE}/sipnet
-        make sipnet
+        curl -L -o sipnet-linux "https://drive.google.com/uc?export=download&id=1sRTFfs9Z9osmMrtLlwVTqxeUDxgVlrEv"
+        chmod +x sipnet-linux
```
Do we make use of `sipnet-linux` while building ?

AritraDey-Dev replied:
#3427 here the binary shared for linux i have updated that.

infotroph replied:
You can see what binary name we use for testing by following the files referenced from line 69 below and finding that `tests/ghaction.sipnet_PostgreSQL.xml` expects to find it at`$GITHUB_WORKSPACE}/sipnet`, so you'll need to update either this line or that file to match each other

infotroph replied:
👍 for now; once https://github.com/PecanProject/sipnet/pull/36 is merged we can come back to update the path to one that will stay up to date as new versions of Sipnet are released.

allgandalf reviewed (commented):
few comments

AritraDey-Dev commented:
@dlebauer @allgandalf  @infotroph I was thinking that how to get the binary file...I have solution please check how much this is possible.After building the binary file based on version or some other parameter,every time it will upload to some location(for now g-drive,or some other location) and then some kind of github Actions which will update the binary file in the sipnet.yml file.What do you think can something like this be done?

AritraDey-Dev committed:
revert: change in ci.yml

AritraDey-Dev commented:
@dlebauer @allgandalf Does this change work?

dlebauer committed:
Merge branch 'develop' into fix_pre-built-sipnet

infotroph committed:
Avoid bottleneck from duplicating settings during sublist extraction

Previous code created extra copies of items on every subsetting operation,
even if the extracted item was not modified.
Profiling with a 100-site MultiSettings showed ~90% of runtime spent here;
comparing x[[1]] in place completely eliminates GCs and runtime goes from
64 sec to 50 ms.

infotroph commented on .github/workflows/sipnet.yml:
```diff
@@ -49,16 +49,15 @@ jobs:
+    - name: Download SIPNET for macOS
+      run: |
+        curl -L -o sipnet-macos "https://drive.google.com/uc?export=download&id=1psNvSw77VcD5QdbSWeGNVQsbCmJokG2W"
+        chmod +x sipnet-macos
```
Don't need to download Mac binary to the Linux check machine

infotroph commented:
@AritraDey-Dev This PR is currently [failing CI checks](https://github.com/PecanProject/pecan/actions/runs/13576219822/job/3795294562) for the same reasons discussed in #3427: The downloaded binary needs to be at the path expected by the test files, and the test needs to be updated to handle retrieving its climate driver file (currently taken from the sipnet repo as `${GITHUB_WORKSPACE}/sipnet/Sites/Niwot/niwot.clim`, which won't work if we're not checking out the repo anymore).

AritraDey-Dev commented:
@infotroph i have a fix for this not sure it will work..will test it and let you know if it works.

AritraDey-Dev committed:
fix: update SIPNET binary download path to match test expectation

AritraDey-Dev committed:
Merge branch 'develop' into fix_pre-built-sipnet

AritraDey-Dev committed:
Merge branch 'develop' into fix_pre-built-sipnet

AritraDey-Dev committed:
fix: update SIPNET binary download path to match test expectation

AritraDey-Dev commented:
@infotroph i have made the changes in test.I have updated the path for the climate driver file in the test configuration to ensure it points to the correct location during the CI process.The CI checks for sipnet is working fine now.

AritraDey-Dev committed:
Merge branch 'develop' into fix_pre-built-sipnet

AritraDey-Dev committed:
Merge branch 'develop' into fix_pre-built-sipnet

AritraDey-Dev commented:
i have made the required changes.can you please look into those.

infotroph commented on .github/workflows/sipnet.yml:
```diff
@@ -49,16 +49,15 @@ jobs:
       run: |
-        cd ${GITHUB_WORKSPACE}/sipnet
-        make sipnet
+        curl -L -o niwot.clim "https://raw.githubusercontent.com/PecanProject/sipnet/refs/heads/master/Sites/Niwot/niwot.clim"
```
Apologies that we moved the ground under you here -- @Alomir and @dlebauer just did a big (and overdue, thank you!) reorganization of the Sipnet repository, and this file has moved. I think for this PR it's enough to change the path:

```suggestion
        curl -L -o niwot.clim "https://raw.githubusercontent.com/PecanProject/sipnet/refs/heads/master/tests/smoke/niwot.clim"
```


Long term there's no guarantee the Sipnet repo will keep providing this file, so we may want to consider providing it somewhere under PEcAn's control, but this works for now.

infotroph commented:
@AritraDey-Dev This PR is now very timely, because the same Sipnet reorganization I mentioned above has also broken our existing checks! If my suggestions above look right to you, let's merge this as our path forward.

infotroph committed:
Merge branch 'develop' into fix_pre-built-sipnet

AritraDey-Dev committed:
fix: correct path of test file

Co-authored-by: Chris Black <chris@ckblack.org>

AritraDey-Dev commented:
@infotroph Thanks for the review! Merging this now as the best path forward.Just to confirm, are there any other potential issues we should address before merging?

infotroph commented:
@AritraDey-Dev No other issues I know of! If we've done everything right, when the currently running CI checks finish we should see that the Sipnet job is passing, at which point I'll press merge (or if not passing, I'll have a look and comment on what's causing the new failure)

AritraDey-Dev commented:
Sounds good! Let's wait for the CI checks to complete. Hopefully, everything passes like last time it did, but if not, i can debug as needed.

AritraDey-Dev commented:
@infotroph The Sipnet checks have passed, so I believe we can proceed with merging.

AritraDey-Dev commented on .github/workflows/sipnet.yml:
```diff
@@ -49,16 +49,15 @@ jobs:
-    - name: install sipnet
+    - name: Download SIPNET for Linux
+      run: |
+        curl -L -o sipnet-linux "https://github.com/PecanProject/sipnet/releases/download/v1.3.0/sipnet-linux-v1.3.0"
```
Hi @infotroph,
I noticed something important in this PR that I thought was worth mentioning.

Currently, the workflow uses a hardcoded SIPNET release version. While the latest release may remain stable for a while, it’s generally a good practice to fetch the release dynamically to ensure the workflow stays up to date automatically with any new versions.

This can be achieved by querying the GitHub API for the latest release tag:
```
     id: get_release
      run: |
        release_tag=$(curl -s https://api.github.com/repos/PecanProject/sipnet/releases/latest | jq -r .tag_name)
        echo "release_tag=$release_tag" >> $GITHUB_OUTPUT
```
Then, you can use the output from this step to construct the download URL dynamically, like so:

`https://github.com/PecanProject/sipnet/releases/download/${{ steps.get_release.outputs.release_tag }}/sipnet-linux-${{ steps.get_release.outputs.release_tag }}`
Let me know your thoughts on this approach and whether we should proceed with integrating it into the project.
